keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,652 @@
+- Protect resources that users marked as in-use
+
+### Goals
+
+- Provide a generic mechanism that prevents a resource from being deleted when it shouldn't be deleted
+
+### Non-Goals
+
+- Provide a specific mechanism to decide which particular objects should be prevented from being deleted when other particular objects exist
+- Provide a generic mechanism that prevents a resource from being updated when it shouldn't be updated
+
+## Proposal
+
+A new field `Liens` to mark the object not to be deleted is introduced in object Metadata.
+Deletion requests for resources with non-empty `Liens` will be blocked by a newly introduced validating admission webhook.
+
+### User Stories (Optional)
+
+#### Story 1
+
+Another controller to protect Secret from deletion watches all Secrets and their potential user objects, like `Pod`, `PersistentVolumes`, and `VolumeSnapshotContent`. Once it find the Secret is used by one of the objects, it updates the Secret's `Liens` and ask this mechanism to block deletion request for the secret.
+
+#### Story 2
+
+A user sets `Liens` to an object to be protected by this mechanism. For example, when a user knows that `Namespace` X shouldn't be erroneously deleted while `Deployment` Y exists for his/her particular use case, the user can set such a protection.
+
+### Notes/Constraints/Caveats (Optional)
+
+- In some cases, users really wan't to delete the objects that are considered to be protected in general.
+For example, immutable Secrets might need to be deleted while it is used by a Pod to update the value.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+Most of the design ideas come from [here](https://github.com/kubernetes/kubernetes/issues/10179).
+Users or controllers can add a string to `Liens` to ask to protect a resource.
+A newly introduced validating admission webhook will reject deletion requests for resources with non-empty `Liens`.
+
+`Liens` is defined as a slice of strings, like `Finalizers`.
+Multiple users or controllers can add their `Liens` for their own purpose, and they can remove their own `Liens` when it is no longer needed.
+Deletion requests of the resource are blocked until the last `Liens` for the resource is removed
+(Just to be clear, the difference between `Liens` and `Finalizers` is that `Liens` blocks the deletion request itself, while `Finalizers` blocks the deletion to be completed).
+
+To make multiple users and controllers share the `Liens` field properly, they should include their identity to the `Liens` string and they should only delete their own `Liens` strings.
```

Comment:
I think there needs to be a debate about the structure. String list is ok, but maybe we can do better. A key-value pair or struct of some sort might be useful for making the owner more obvious.

Reply:
I'm not so convinced here - consistency with finalizers is nice, and is it really confusing?

Reply:
Thinking a bit more about that - clearly a struct would be more extensible.
But in this particular case, I'm not convinced it's actually an advantage.

I already commented below on scalability implications if we decide to store Lien per object - I don't think it's a good solution. And making this extensible would be definition be prone to those kind of regressions.

So my personal preference is actually towards a simple string (which BTW as Tim pointed out again would be consistent with Finalizers which is pretty nice).

Reply:
Change back to slice-of-strings Liens and deleted this line.